v1/remote/image_test.go

```diff
@@ -0,0 +1,292 @@
+		ref           string
+		responseBody  []byte
+		contentDigest string
+		wantErr       bool
+	}{{
+		// Normal pull, by tag.
+		ref:           "latest",
+		responseBody:  mustRawManifest(t, img),
+		contentDigest: mustDigest(t, img).String(),
+		wantErr:       false,
+	}, {
+		// Normal pull, by digest.
+		ref:           mustDigest(t, img).String(),
+		responseBody:  mustRawManifest(t, img),
+		contentDigest: mustDigest(t, img).String(),
+		wantErr:       false,
+	}, {
+		// Right content-digest, wrong body, by tag.
+		ref:           "latest",
+		responseBody:  []byte("not even json"),
+		contentDigest: mustDigest(t, img).String(),
+		wantErr:       true,
+	}, {
+		// Right content-digest, wrong body, by digest.
+		ref:           mustDigest(t, img).String(),
+		responseBody:  []byte("not even json"),
+		contentDigest: mustDigest(t, img).String(),
+		wantErr:       true,
+	}, {
+		// Right body, wrong content-digest, by tag.
+		ref:           "latest",
+		responseBody:  mustRawManifest(t, img),
+		contentDigest: bogusDigest,
+		wantErr:       true,
+	}, {
+		// Right body, wrong content-digest, by digest.
+		// NB: This succeeds! We don't care what the registry thinks.
+		ref:           mustDigest(t, img).String(),
+		responseBody:  mustRawManifest(t, img),
+		contentDigest: bogusDigest,
+		wantErr:       false,
+	}, {
+		// Nothing matches anything.
+		ref:           "latest",
+		responseBody:  []byte("everything is wrong with this"),
+		contentDigest: bogusDigest,
+		wantErr:       true,
+	}}
+
+	for _, tc := range cases {
+		manifestPath := fmt.Sprintf("/v2/%s/manifests/%s", expectedRepo, tc.ref)
+		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			switch r.URL.Path {
+			case manifestPath:
+				if r.Method != http.MethodGet {
+					t.Errorf("Method; got %v, want %v", r.Method, http.MethodGet)
+				}
+
+				w.Header().Set("Docker-Content-Digest", tc.contentDigest)
+				w.Write(tc.responseBody)
+			default:
+				t.Fatalf("Unexpected path: %v", r.URL.Path)
+			}
+		}))
+		defer server.Close()
+		u, err := url.Parse(server.URL)
+		if err != nil {
+			t.Fatalf("url.Parse(%v) = %v", server.URL, err)
+		}
+
+		ref, err := newReference(u.Host, expectedRepo, tc.ref)
+		if err != nil {
+			t.Fatalf("url.Parse(%v, %v, %v) = %v", u.Host, expectedRepo, tc.ref, err)
+		}
+
+		rmt := remoteImage{
+			ref:    ref,
+			client: http.DefaultClient,
+		}
+
+		if _, err := rmt.RawManifest(); (err != nil) != tc.wantErr {
+			t.Error("RawManifest() wrong error: %v, want %v: %v", (err != nil), tc.wantErr, err)
+		}
+	}
+}
+
+func TestRawManifestNotFound(t *testing.T) {
+	expectedRepo := "foo/bar"
+	manifestPath := fmt.Sprintf("/v2/%s/manifests/latest", expectedRepo)
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		switch r.URL.Path {
+		case manifestPath:
+			if r.Method != http.MethodGet {
+				t.Errorf("Method; got %v, want %v", r.Method, http.MethodGet)
+			}
```

Comment:
I noticed that we're not checking `Accept` headers in any of the tests.  Not sure how we want to handle this.

Reply:
Added TestAcceptHeaders. That will get more complicated when I write the manifest list handling.